twine/settings.py

```diff
@@ -263,6 +265,13 @@ def register_argparse_arguments(parser: argparse.ArgumentParser) -> None:
             action="store_true",
             help="Disable the progress bar.",
         )
+        parser.add_argument(
+            "--no-color",
+            default=False,
+            required=False,
+            action="store_true",
+            help="Disable colored output.",
+        )
```

Comment:
I think the `--no-color` argument and the `no_color` attribute should be removed from this class, and possibly moved to a utility class, e.g.:

```python
# cli.py
def build_command_parser(command: str) -> argparse.ArgumentParser:
    parser = cli.argparse.ArgumentParser(prog=f"twine {command}")
    parser.add_argument(
        "--no-color",
        # ...
    )
    return parser


# commands/check.py
def main(args: List[str]) -> bool:
    parser = cli.build_command_parser("check")
    # ...


# commands/upload.py
def main(args: List[str]) -> None:
    parser = cli.build_command_parser("upload")
    settings.Settings.register_argparse_arguments(parser)
    # ...
```

I'm not sure if `cli.py` is the right place for that.

Reply:
I updated my comment above to allow `--no-color` to *also*  work when `twine` is run without a command.

Another half-baked thought: in `cli.py`, `parser` could be a module attribute, initialized by `build_parser`. That would make it accessible globally, which I think would address my concern in https://github.com/pypa/twine/pull/649#discussion_r439812534.

Or, we could rework it entirely, to use the `argparse` support for [sub-commands](https://docs.python.org/3/library/argparse.html#sub-commands).

Reply:
@bhrutledge Any idea which you'd prefer? I'm not sure @callmecampos has enough information to move forward here.

Reply:
I'm admittedly a little stuck on this. I played around with implementing some of my ideas above, but ran into some complexities that I don't have time to elaborate on at the moment.

Re: using `Settings` as suggested in https://github.com/pypa/twine/issues/659#issuecomment-643221227, that comment only refers to `upload` and `register`, so it's not accurate that *all* of Twine's commands use `Settings`. If we want to use that, it seems like we'd have to add conditional `argparse` behavior based on the current command. 

Re: reworking to use `argparse` sub-commands, that feels out of scope for this PR.

Questions:

- Why do we need `--no-color`? I believe I saw it discussed elsewhere, but can't find it. If colorama automatically disables color when not printing to an interactive terminal, could that be sufficient for now?

- If we need `--no-color`, does it need to be on the base `twine` command? I tweaked my code sample above to allow that, but after playing around with it, I don't think it does. If not, I'd revert to my earlier suggestion of `build_command_parser` (visible in the comment revision history).

- If we need `--no-color`, how do we propagate that setting (and future global settings) from the argument parsing in commands to the top-level handlers like `__main__` and `cli.dispatch`? I'd like to avoid the current approach of re-parsing the arguments, because that feels brittle.

Reply:
> Why do we need --no-color? I believe I saw it discussed elsewhere, but can't find it. If colorama automatically disables color when not printing to an interactive terminal, could that be sufficient for now?

I don't think we know of a specific use case where this would be desirable but I'm sure it exists. I wouldn't expect that Colorama works 100% perfectly in every terminal/shell/platform/etc. I think it's generally good practice to add an off switch when adding an optional feature like this.

> If we need --no-color, does it need to be on the base twine command? I tweaked my code sample above to allow that, but after playing around with it, I don't think it does. If not, I'd revert to my earlier suggestion of build_command_parser (visible in the comment revision history).

It seems like the alternative is to either add it to every command manually. I'm not sure that's the best solution. Ideally flags shared by all commands could just be defined once somewhere.

> If we need --no-color, how do we propagate that setting (and future global settings) from the argument parsing in commands to the top-level handlers like __main__ and cli.dispatch? I'd like to avoid the current approach of re-parsing the arguments, because that feels brittle.

I agree this is a problem. I think https://github.com/pypa/twine/issues/659 would resolve this but it seems like that's not ideal because it might break some third-parties that are using these unsupported internal APIs. I don't have a better solution, though.